src/cachingreader.cpp

```diff
@@ -258,199 +266,161 @@ void CachingReader::process() {
+                pChunk->state = Chunk::READ;
+            } else if (status.status == CHUNK_READ_EOF) {
+                Chunk* pChunk = status.chunk;
+                if (pChunk == NULL) {
+                    qDebug() << "ERROR: status.chunk is NULL in CHUNK_READ_EOF ReaderStatusUpdate. Ignoring update.";
+                    continue;
+                }
+                DEBUG_ASSERT(pChunk->state == Chunk::READ_IN_PROGRESS);
+                freeChunk(pChunk);
+            } else if (status.status == CHUNK_READ_INVALID) {
+                qDebug() << "WARNING: READER THREAD RECEIVED INVALID CHUNK READ";
+                Chunk* pChunk = status.chunk;
+                if (pChunk == NULL) {
+                    qDebug() << "ERROR: status.chunk is NULL in CHUNK_READ_INVALID ReaderStatusUpdate. Ignoring update.";
+                    continue;
+                }
+                DEBUG_ASSERT(pChunk->state == Chunk::READ_IN_PROGRESS);
+                freeChunk(pChunk);
             }
-            DEBUG_ASSERT(pChunk->state == Chunk::READ_IN_PROGRESS);
-            freeChunk(pChunk);
         }
     }
 }
 
-int CachingReader::read(int sample, int num_samples, CSAMPLE* buffer) {
+int CachingReader::read(int sample, int numSamples, CSAMPLE* buffer) {
     // Check for bad inputs
-    DEBUG_ASSERT_AND_HANDLE(sample % 2 == 0) {
+    DEBUG_ASSERT_AND_HANDLE(sample % CachingReaderWorker::kChunkChannels == 0) {
         // This problem is easy to fix, but this type of call should be
         // complained about loudly.
         --sample;
     }
-    DEBUG_ASSERT_AND_HANDLE(num_samples % 2 == 0) {
-        --num_samples;
+    DEBUG_ASSERT_AND_HANDLE(numSamples % CachingReaderWorker::kChunkChannels == 0) {
+        --numSamples;
     }
-    if (num_samples < 0 || !buffer) {
+    if (numSamples < 0 || !buffer) {
         QString temp = QString("Sample = %1").arg(sample);
         qDebug() << "CachingReader::read() invalid arguments sample:" << sample
-                 << "num_samples:" << num_samples << "buffer:" << buffer;
+                 << "numSamples:" << numSamples << "buffer:" << buffer;
         return 0;
     }
 
     // If asked to read 0 samples, don't do anything. (this is a perfectly
     // reasonable request that happens sometimes. If no track is loaded, don't
     // do anything.
-    if (num_samples == 0 || m_readerStatus != TRACK_LOADED) {
+    if (numSamples == 0 || m_readerStatus != TRACK_LOADED) {
         return 0;
     }
 
     // Process messages from the reader thread.
     process();
 
-    int samples_remaining = num_samples;
+    SINT samplesRead = 0;
 
+    SINT frameIndex = samples2frames(sample);
+    SINT numFrames = samples2frames(numSamples);
 
     // TODO: is it possible to move this code out of caching reader
     // and into enginebuffer?  It doesn't quite make sense here, although
     // it makes preroll completely transparent to the rest of the code
     // if we're in preroll...
```

Comment:
It is not the first time I am struggling with this comment. I like the current solution, because it simplifies the code elsewhere. If you think the same, we should remove this comment.